### PR TITLE
Merge into the inMemoryDB

### PR DESCRIPTION
**Pull request by cp5555**
(1) Add the error log and integrate it into ModSecurity
(2) Fix bug in SHM_create()

### CODE REVIEW

File: apache2/persist_dbm.c [hunk 1]
```diff
@@ -320,7 +321,9 @@ static apr_table_t *collection_retrieve_ex(int db_option, void *existing_dbm, mo
                     goto cleanup;
                 }
             }
-            rc2 = AGMDB_delete(ag_dbm, col_key, col_key_len);
+            rc = AGMDB_delete(ag_dbm, col_key, col_key_len);
+            if(rc != AGMDB_SUCCESS)
```
Comment (azhao155) [inline apache2/persist_dbm.c]: why do we move the this code

File: apache2/ag_mdb/ag_mdb.cpp [hunk 2]
```diff
@@ -956,36 +970,33 @@ int AGMDB_get(struct agmdb_handler *dbm, const char* key, int key_len, char* buf
...
     // Hash the key, and search the key in the corresponding hash linklist
     entry_id = AGMDB_getEntry(shm_base, key, key_len);
     if(entry_id == AGMDB_INVALID_INDEX) // Cannot find this key, need a new entry
-        rc = AGMDB_setSpareEntry(shm_base, key, key_len, value, value_len);
+        return AGMDB_setSpareEntry(shm_base, key, key_len, value, value_len);
     else 
-        rc = AGMDB_setEntry(shm_base, entry_id, key, key_len, value, value_len);
-    if(rc != AGMDB_SUCCESS)
-        return rc;
-    return AGMDB_SUCCESS;
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp:988]: why do we write this extra condition before?

Comment (cp5555) [reply apache2/ag_mdb/ag_mdb.cpp:988]: Just for optimization

File: apache2/persist_dbm.c (hunk 1 above)
Comment (cp5555) [reply apache2/persist_dbm.c]: I think rc2 is not clear, so I just change it to tmp_val_len

File: apache2/persist_dbm.c [hunk 3]
```diff
@@ -639,7 +639,7 @@ static int collection_store_ex(int db_option, modsec_rec *msr, apr_table_t *col)
 #ifdef GLOBAL_COLLECTION_LOCK
             apr_global_mutex_unlock(msr->modsecurity->dbm_lock);
 #endif
-            msr_log(msr, 1, "[ERRNO]collection_store_ex_origin: Failed to access DBM file \"%s\": %s", log_escape(msr->mp, dbm_filename),
```
Comment (azhao155) [inline apache2/persist_dbm.c:641]: why do we delete all the ERRNO in the log msg?

File: apache2/ag_mdb/ag_mdb_external.h [hunk 4]
```diff
@@ -66,28 +122,28 @@ struct agmdb_handler{
...
- ** return: AGMDB_SUCCESS if successfully created or AGMDB_FAIL if failed.
+ ** return: AGMDB_SUCCESS if successfully created or AGMDB_ERROR if failed.
  */
 int AGMDB_getSharedLock(struct agmdb_handler *dbm);
 
 /**
  ** Get a exclusive lock for read and write.
  ** @param dbm: the database you want to lock.
- ** return: AGMDB_SUCCESS if successfully created or AGMDB_FAIL if failed.
+ ** return: AGMDB_SUCCESS if successfully created or AGMDB_ERROR if failed.
```
Comment (cp5555) [inline apache2/ag_mdb/ag_mdb_external.h]: test

File: apache2/ag_mdb/ag_mdb.cpp [hunk 5]
```diff
@@ -660,36 +670,40 @@ int AGMDB_getExclusiveLock(struct agmdb_handler *dbm) {
...
     db_lock = (struct agmdb_lock *)&(dbm->db_lock);
     
-    if(Lock_V(db_lock, SEM_ID_READ, SEM_READ_INITVAL) == AGMDB_FAIL)
-            return AGMDB_FAIL;
-        
-    if(Lock_V(db_lock, SEM_ID_WRITE, 1) == AGMDB_FAIL)
-        return AGMDB_FAIL;
+    rc = Lock_V(db_lock, SEM_ID_READ, SEM_READ_INITVAL);
+    if (rc != AGMDB_SUCCESS)
+        return AGMDB_ERROR;
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp:828]: why don't we return the specific error code here intead of AGMDB_ERROR

File: apache2/ag_mdb/ag_mdb.cpp [hunk 6]
```diff
@@ -1077,15 +1077,121 @@ int AGMDB_removeStale(struct agmdb_handler *dbm) {
     return AGMDB_SUCCESS;
 }
 
-const char* AGMDB_getErrorInfo(int error_code){
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp]: why this function got added and then deleted

File: apache2/persist_dbm.c [hunk 7]
```diff
@@ -906,7 +907,9 @@ static int collections_remove_stale_ex(int db_option, modsec_rec *msr, const cha
         ag_dbm = dcfg_searchAGMDBhandler(col_name, (struct agmdb_handle_entry*)(root_dcfg->agmdb_handles));
         if(ag_dbm == NULL)
             return 1;
-        return AGMDB_removeStale(ag_dbm);
+        rc2 = AGMDB_removeStale(ag_dbm);
```
Comment (azhao155) [inline apache2/persist_dbm.c:943]: why we use rc2 and rc for the return value, should we have more consistent coding style?

File: apache2/persist_dbm.c [hunk 8]
```diff
@@ -14,17 +14,11 @@
 
 #include "persist_dbm.h"
 #include "apr_sdbm.h"
+#ifdef MEMORY_DATABASE_ENABLE
 #include "ag_mdb_external.h"
-#include "errno.h"
```
Comment (azhao155) [inline apache2/persist_dbm.c:18]: why delete these code in the compiling change?

File: apache2/ag_mdb/ag_mdb.cpp [hunk 9]
```diff
@@ -57,36 +56,91 @@ int Lock_create(struct agmdb_lock *new_lock, const char* lock_name, int lock_nam
         return AGMDB_ERROR_LOCK_WIN_NAME_INVALID_STRING;
     sprintf_s(read_lock_name, AGMDB_MAX_NAME_LEN, "DB_%s_LOCK_READ", lock_name);
     sprintf_s(write_lock_name, AGMDB_MAX_NAME_LEN, "DB_%s_LOCK_WRITE", lock_name);
-    new_mutex = CreateMutex(
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp]: why do we change from createMutex, OpenMutex to just createMutex, try to get more context

File: apache2/ag_mdb/ag_mdb.cpp [hunk 10]
```diff
@@ -321,6 +383,54 @@ int SHM_create(PTR_VOID* new_shm_base, const char* db_name, int db_name_length,
...
+    return AGMDB_SUCCESS;
+}
+
+/**
+ ** Close the shared memory of AG Memory Database
+ ** @param dbm: The handler of the database.
+ ** return: AGMDB_SUCCESS if successfully close shared memory,
+ **         AGMDB_ERROR if failed. 
+ */
+int SHM_close(struct agmdb_handler* dbm) {
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp:411]: is there a reason for us to add this function, so under what circumstance, it needs the process to de-attach?

File: apache2/ag_mdb/ag_mdb.cpp [hunk 11]
```diff
@@ -57,36 +56,91 @@ int Lock_create(struct agmdb_lock *new_lock, const char* lock_name, int lock_nam
...
 }
 
+/**
+ ** Destroy the lock of AG Memory Database 
+ ** @param db_lock: The agmdb_lock sturcture  
+ ** return: AGMDB_SUCCESS if successfully destroy the lock, 
+ **         AGMDB_ERROR if failed.
+ */
+int Lock_destroy(struct agmdb_lock *db_lock) {
+    /* Windows doesn't do anything */
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp:103]: so in windows, if the process exit, the lock will be automatically destroyed by the OS?

File: apache2/ag_mdb/ag_mdb.cpp [hunk 12]
```diff
@@ -57,36 +56,91 @@ int Lock_create(struct agmdb_lock *new_lock, const char* lock_name, int lock_nam
...
+}
+
+/**
+ ** Close the lock of AG Memory Database 
+ ** @param db_lock: The agmdb_lock sturcture  
+ ** return: AGMDB_SUCCESS if successfully close the lock,
+ **         AGMDB_ERROR if failed. 
+ */
+int Lock_close(struct agmdb_lock *db_lock) {
+    /* Linux doesn't do anything */
```
Comment (azhao155) [inline apache2/ag_mdb/ag_mdb.cpp:120]: So in Linux, do we need to set the handle to NULL?

File: apache2/ag_mdb/ag_mdb.cpp (hunk 5 above)
Comment (cp5555) [reply apache2/ag_mdb/ag_mdb.cpp:828]: Will revise it

File: apache2/persist_dbm.c (hunk 8 above)
Comment (cp5555) [reply apache2/persist_dbm.c:18]: I have moved it into ag_mdb_external.h

File: apache2/ag_mdb/ag_mdb.cpp (hunk 12 above)
Comment (cp5555) [reply apache2/ag_mdb/ag_mdb.cpp:120]: I have rewrite this function. We can skip it.

**Pull request merged**